Source/Core/DolphinQt/RenderWidget.cpp

```diff
@@ -192,7 +201,35 @@ bool RenderWidget::event(QEvent* event)
     }
     break;
   case QEvent::WinIdChange:
-    emit HandleChanged(reinterpret_cast<void*>(winId()));
+    if (m_wsi_type != WindowSystemType::Wayland)
+        emit HandleChanged(reinterpret_cast<void*>(winId()));
+    break;
+  case QEvent::PlatformSurface:
+    // On Wayland, the winId is irrelevant for rebinding the renderer surface.
+    // The PlatformSurfaceEvent provides a means to synchronously handle wl_surface creations and
+    // destroys.
+    if (m_wsi_type == WindowSystemType::Wayland)
+    {
+      auto* surface_event = static_cast<QPlatformSurfaceEvent*>(event);
+      switch (surface_event->surfaceEventType())
+      {
+      case QPlatformSurfaceEvent::SurfaceCreated:
+        // The actual surface creation handler must occur within its own event invocation to avoid
+        // wayland-egl deadlocks.
+        QGuiApplication::postEvent(this, new QEvent(QEvent::User));
+        break;
+      case QPlatformSurfaceEvent::SurfaceAboutToBeDestroyed:
+        emit SurfaceAboutToBeDestroyed();
+        break;
+      default:
+        break;
+      }
+    }
+    break;
+  case QEvent::User:
+    // Followup from QPlatformSurfaceEvent::SurfaceCreated
+    emit SurfaceCreated(QGuiApplication::platformNativeInterface()->nativeResourceForWindow(
+        "surface", windowHandle()));
```

Comment:
This took me a couple days to figure out x3

WinIdChange is effectively pointless on wayland. You can't actually use the value to get a wl_surface. I think it's private to QtWayland.

I've added an alterative API for handle changes to block the EmuThread in between surface destruction and creation so QtWayland can safely destroy and recreate the wl_surface.